context/app/static/js/components/detailPage/MetadataSection/utils.ts

```diff
@@ -0,0 +1,152 @@
+import { nodeIcons } from 'js/components/detailPage/DatasetRelationships/nodeTypes';
+import { TableRows } from 'js/components/detailPage/MetadataSection/MetadataSection';
+import { Dataset, Donor, ESEntityType, Sample, isDataset, isSample } from 'js/components/types';
+import { getMetadata } from 'js/helpers/metadata';
+import { ProcessedDatasetInfo } from 'js/pages/Dataset/hooks';
+import { MUIIcon, entityIconMap } from 'js/shared-styles/icons/entityIconMap';
+
+function getDescription(field: string, metadataFieldDescriptions: Record<string, string> | Record<string, never>) {
+  const [prefix, stem] = field.split('.');
+  if (!stem) {
+    return metadataFieldDescriptions?.[field];
+  }
+  const description = metadataFieldDescriptions?.[stem];
+  if (!description) {
+    return undefined;
+  }
+  if (prefix === 'donor') {
+    return `For the original donor: ${metadataFieldDescriptions?.[stem]}`;
+  }
+  if (prefix === 'sample') {
+    return `For the original sample: ${metadataFieldDescriptions?.[stem]}`;
+  }
+  throw new Error(`Unrecognized metadata field prefix: ${prefix}`);
+}
+
+function buildTableData(
+  tableData: Record<string, string | object | unknown[]>,
+  metadataFieldDescriptions: Record<string, string> | Record<string, never>,
+  extraValues: Record<string, string> = {},
+) {
+  return (
+    Object.entries(tableData)
+      // Filter out nested objects, like nested "metadata" for Samples...
+      // but allow arrays. Remember, in JS: typeof [] === 'object'
+      .filter((entry) => typeof entry[1] !== 'object' || Array.isArray(entry[1]))
+      // Filter out fields from TSV that aren't really metadata:
+      .filter((entry) => !['contributors_path', 'antibodies_path', 'version'].includes(entry[0]))
+      .map((entry) => ({
+        ...extraValues,
+        key: entry[0],
+        // eslint-disable-next-line @typescript-eslint/no-base-to-string
+        value: Array.isArray(entry[1]) ? entry[1].join(', ') : entry[1].toString(),
+        description: getDescription(entry[0], metadataFieldDescriptions),
+      }))
+  );
+}
+
+function getEntityIcon(entity: { entity_type: ESEntityType; is_component?: boolean; processing?: string }) {
+  if (isDataset(entity)) {
+    if (entity.is_component) {
+      return nodeIcons.componentDataset;
+    }
+    if (entity.processing === 'processed') {
+      return nodeIcons.processedDataset;
+    }
+    return nodeIcons.primaryDataset;
+  }
+  return entityIconMap[entity.entity_type];
+}
+
+export interface TableEntity {
+  uuid: string;
+  label: string;
+  icon: MUIIcon;
+  tableRows: TableRows;
+  entity_type: ESEntityType;
+  hubmap_id: string;
+}
+
+interface sortEntitiesProps {
+  tableEntities: TableEntity[];
+  uuid: string;
+}
+
+function sortEntities({ tableEntities, uuid }: sortEntitiesProps) {
+  return tableEntities.sort((a, b) => {
```

Comment:
To avoid mutating the tableEntities object:

```suggestion
  return [...tableEntities].sort((a, b) => {
```